shipyard_blueprints/dbt/shipyard_dbt/dbt.py

```diff
@@ -18,21 +28,149 @@ def __init__(self, access_token: str, account_id: str):
                 return 0
             else:
-                self.logger.error("Could not connect to DBT")
+                logger.error("Could not connect to DBT")
                 return 1
         except Exception as e:
-            self.logger.error("Could not connect to DBT")
+            logger.error(f"Could not connect to DBT due to {e}")
             return 1
+
+    def _determine_connection_status(self, response):
+        status_code = response["status"]["code"]
+        user_message = response["status"]["user_message"]
+        if status_code == 401:
+            if "Invalid token" in user_message:
+                raise ExitCodeException(
+                    f"The Service Token provided was invalid. Check to make sure there are no typos or "
+                    f"preceding/trailing spaces. dbt API says: {user_message}",
+                    self.EXIT_CODE_INVALID_CREDENTIALS,
+                )
+            else:
+                raise ExitCodeException(
+                    f"An unknown error occurred with a status code of {status_code}. dbt API says: {user_message}",
+                    self.EXIT_CODE_UNKNOWN_ERROR,
+                )
+        if status_code == 404:
+            raise ExitCodeException(
+                f"The Account ID, Job ID, or Run ID provided was either invalid or your API Key doesn't"
+                f"have access to it. Check to make sure there are no typos or preceding/trailing spaces. "
+                f"dbt API says: {user_message}",
+                self.EXIT_CODE_INVALID_RESOURCE,
+            )
+
+    def get_run_details(self, run_id):
+        logger.info(f"Grabbing run details for run {run_id}.")
+
+        response = self._request(
+            url=f"{self.account_url}/runs/{run_id}/",
+            params={"include_related": "['run_steps','debug_logs']"},
+        )
+        logger.debug(f"Response from the API: \n{response}")
+        return response
+
+    def get_artifact_details(self, run_id):
+        logger.info(f"Grabbing artifact details for run {run_id}")
+        return self._request(f"{self.account_url}/runs/{run_id}/artifacts/")
+
+    def download_artifact(self, run_id, artifact_name, destination_folder):
+        get_artifact_details_url = (
+            f"{self.account_url}/runs/{run_id}/artifacts/{artifact_name}"
+        )
+        artifact_file_name = os.path.basename(artifact_name)
+        artifact_folder = os.path.dirname(artifact_name)
+
+        destination_fullpath = os.path.join(destination_folder, artifact_folder)
+        os.makedirs(destination_fullpath, exist_ok=True)
+
+        filename = os.path.join(destination_fullpath, artifact_file_name)
+
+        try:
+            with requests.get(
+                    get_artifact_details_url, headers=self.headers, stream=True
+            ) as r:
+                r.raise_for_status()
+                with open(filename, "wb") as f:
+                    for chunk in r.iter_content(chunk_size=(16 * 1024 * 1024)):
```

Comment:
Not to be pedantic, but define the chunk_size as a constant so we don't have magic numbers